### ISSUE TITLE
Use manual remove for PCI device instead of Bbswitch?

### ISSUE THREAD

**Issue by bomiyr** [openSUSE/SUSEPrime#74]
Hi, It's me again :smile:  
In [this issue](https://github.com/openSUSE/SUSEPrime/issues/73) we found out that bbswitch is not working on my laptop. TLDR about my investigation is in [this comment](https://github.com/openSUSE/SUSEPrime/issues/73#issuecomment-974553332).

Finally I was able to disable discrete GPU with just `echo 1 | sudo tee /sys/bus/pci/devices/0000\:01\:00.0/remove`.

And I'm started thinking, that maybe this is the right way for disabling the device? In this case we don't need to have dependency for bbswitch (which seems to be almost dead, if you look at it's repo). But I don't have the right expertise on the topic to see the whole picture. 

So what do you think, is it possible to integrate such solution into suse-prime itself, or there are some hidden caveats in such approach?

**Comment by sndirsch** [comment]
Thanks for the report. Indeed I could disable the NVIDIA GPU that way. :-)

linux:/home/tux # modprobe nvidia
modprobe: ERROR: could not insert 'nvidia': No such device

linux:/home/tux # dmesg
[ 5172.008516] nvidia-nvlink: Nvlink Core is being initialized, major device number 236
[ 5172.008832] NVRM: No NVIDIA GPU found.
[ 5172.009523] nvidia-nvlink: Unregistered the Nvlink Core, major device number 236

Unfortunately I was not able to reenable it. :-( I've tried 

linux:/home/tux # echo 1 | tee /sys/bus/pci/devices/0000\:01\:00.0/rescan
tee: '/sys/bus/pci/devices/0000:01:00.0/rescan': No such file or directory
1

linux:/home/tux # echo 1  /sys/bus/pci/devices/0000\:01\:00.0/rescan
1 /sys/bus/pci/devices/0000:01:00.0/rescan

linux:/home/tux #  modprobe nvidia
modprobe: ERROR: could not insert 'nvidia': No such device
linux:/home/tux # dmesg
[ 5424.197577] nvidia-nvlink: Nvlink Core is being initialized, major device number 236
[ 5424.197879] NVRM: No NVIDIA GPU found.
[ 5424.198315] nvidia-nvlink: Unregistered the Nvlink Core, major device number 236

**Comment by bomiyr** [comment]
I was able to reenable card with
’echo 1 | sudo tee /sys/bus/pci/rescan’

**Comment by sndirsch** [comment]
OMG. ;-) Indeed now it works for me. :-) Even easier.

# disable nVidia GPU
modprobe -r nvidia_drm nvidia_modeset nvidia_uvm nvidia
echo 1 > /sys/bus/pci/devices/0000\:00\:01/remove

# reenable nVidia GPU
echo 1 > /sys/bus/pci/rescan

**Comment by sndirsch** [comment]
Adding our bbswitch expert. @simopil What do you think? Should we try to get rid of bbswitch? Looks like it would be possible.

**Comment by bomiyr** [comment]
The only problem I see is that _rescan_ command is not exclusive to GPU and will add any previously removed PCI device. And vice versa, calling _rescan_ by user or by some hardware-info app will silently enable GPU...

**Comment by sndirsch** [comment]
Yeah. Valid points. And there might be more reasons, why bbswitch kernel module has been written ...

**Comment by simopil** [comment]
I tried removing device, it disappeared from system but leds on my laptop shows that nvidia card is still powered on. 
This is because powering off gpu is done via ACPI calls and not via unbinding device. You can power it off via [acpi_call](https://github.com/mkottman/acpi_call) with correct [call](https://github.com/mkottman/acpi_call/blob/master/examples/turn_off_gpu.sh) for your platform. 
bbswitch can find a suitable call for your system automatically and perform it without acpi_call module.
BTW bbswitch module is mantained despite main repo is very old, one year ago it was broken due kernel libraries changes and it was fixed in openSUSE

**Comment by sndirsch** [comment]
Thanks for your input @simopil !
